pricegraph/src/graph/shortest_paths.rs

```diff
@@ -2,51 +2,65 @@
 use petgraph::algo::FloatMeasure;
 use petgraph::visit::{
     Data, EdgeRef, GraphBase, IntoEdges, IntoNodeIdentifiers, NodeCount, NodeIndexable,
 };
+use unbounded::Unbounded;
+
+mod unbounded;
 
 /// A vector associating to each node index a distance from the source.
-type Distance<G> = Vec<<G as Data>::EdgeWeight>;
-
-/// Stores the information needed to manage the predecessor list.
-/// For each node index, this type contains its predecessor node and
-/// distance in the graph from a source node.
-struct PredecessorStore<G: GraphBase + Data> {
-    predecessors: Vec<Option<G::NodeId>>,
-    distance: Distance<G>,
-}
+type Distances<G> = Vec<<G as Data>::EdgeWeight>;
 
-impl<G: GraphBase + Data> PredecessorStore<G> {
-    fn distance(&self, node_index: usize) -> &G::EdgeWeight {
-        &self.distance[node_index]
-    }
+/// A Vector associating to each node index the preceding node on the shortest path
+type PredecessorVec<G> = Vec<Option<<G as GraphBase>::NodeId>>;
 
-    fn update_distance(&mut self, node_index: usize, updated_distance: G::EdgeWeight) {
-        self.distance[node_index] = updated_distance;
-    }
+trait PredecessorStoring<G: GraphBase + Data> {
+    /// Returns the current distance of a node from the source.
+    fn distance(&self, node_index: usize) -> G::EdgeWeight;
 
-    fn update_predecessor(&mut self, node_index: usize, updated_predecessor: Option<G::NodeId>) {
-        self.predecessors[node_index] = updated_predecessor;
-    }
+    /// Updates the distance of a node from the source.
+    fn update_distance(&mut self, node_index: usize, updated_distance: G::EdgeWeight);
+
+    /// Updates the predecessor of a node.
+    fn update_predecessor(&mut self, node_index: usize, updated_predecessor: Option<G::NodeId>);
+
+    /// Returns shortest path from source to destination node, if a path exists.
+    fn path_to(&self, source: G::NodeId, dest: G::NodeId, graph: G) -> Option<Path<G::NodeId>>;
+
+    /// Lists all nodes that can be reached from the source (excluding the source itself).
+    fn connected_nodes(&self, graph: G) -> Vec<G::NodeId>;
+
+    /// Callback for any work that needs to be done before each pass of the bellman-ford algorithm
+    fn prepare_next_relaxation_step(&mut self);
+
+    /// Checks for negative weight cycle and, if any is found, creates loop in
+    /// the predecessor store
+    fn mark_cycle(&mut self, graph: G) -> Option<G::NodeId>;
+
+    /// Returns a negative cycle from the starting node, if it exists.
+    fn find_cycle(&mut self, search_start: G::NodeId, graph: G)
+        -> Option<NegativeCycle<G::NodeId>>;
 }
 
 /// Structure that can be used to derive the shorthest path from a source to any
 /// reachable destination in the graph.
-pub struct ShortestPathGraph<G: GraphBase + Data> {
+pub struct ShortestPathGraph<'a, G: Data> {
     graph: G,
-    predecessor_store: PredecessorStore<G>,
+    predecessor_store: Box<dyn PredecessorStoring<G> + 'a>,
```

Comment:
This could slow down the compiled coded because the compiler might be unable to see through the trait object (`dyn`) which prevents inlining. To fix that, it should be a type parameter `<T: PredecessorStoring>`. But then I can see that this would make the code wants to use ShortestPathGraph harder because we would have to turn it into a trait and a return a `Box<dyn ShortestPathGraph>` when creating it.
Maybe performance impact is very small and we can still do this change later so this isn't a blocker.

Reply:
I can benchmark it...

Reply:
Performance for this PR regressed much more than what @fedgiac saw in his, so we might want to take a look at improving this:

||This PR|#1368
|---|---|---
|Pricegraph::read | +74.929%|+25.719%
|Pricegraph::transitive_orderbook/5298183|+75.250%|+25.110%
|Pricegraph::estimate_limit_price/100000000000000000000 |+16.503|+14.982%
|Pricegraph::estimate_limit_price/1000000000000000000000|+21.946%|+19.747%
|Pricegraph::order_for_limit_price/100|+25.230%|+10.584%

I took a crack at implementing the flattending but am not able to resolve the type errors I'm seeing (made a commit [here](https://github.com/gnosis/dex-services/commit/af88a0d089a9802c316c99b02fd0a8dde41c7f46) with my current state)

Reply:
I now have a compilable version of the flattened code (needed to make `ShortestPathGraph::empty/new` free functions), but am still seeing the same performance as in this PR.

My changes are on top of the next PR and committed in https://github.com/gnosis/dex-services/commit/4a48742cf71dcc4ebceb26a88a27732d1ef3fd1d

Reply:
@e00E helped me figure out a way how to get rid of the dynamic dispatch inside the `bellman_ford` method and thus reducing the overhead to similar (even slightly better) levels than the original PR (#1368).

This does however affect a bunch of callsites of the ShortestPath Graph so I would like to do it in a separate PR on top of #1495 if that is ok with you...